docs/console-access.md

```diff
@@ -0,0 +1,340 @@
+Console access for virtual machines
+===================================
+
+There are two main ways tenant users may interact with virtual machines they
+are running.
+
+ * Guest OS hosted network services. This covers SSH server, web apps like
+   cockpit, Windows remote desktop. The set of services available varies
+   according to what OS is installed, and can only be accessed if the guest
+   is running "normally".
+
+ * Host OS hosted network services. This covers VNC, SPICE, and serial port
+   console services exposed over the host network by the QEMU emulator. These
+   are available as long as the virtual machine is powered on, regardless of
+   whether the guest OS is running "normally" or not. The set of services
+   available is the same regardless of what kind of guest OS is installed.
+
+Access to guest OS hosted network services does not require any special work
+by kubevirt aside from ensuring the guest has a working network connection to
+the tenant. This document thus considers how to expose the host OS hosted
+network services to tenant users.
+
+
+Proxied vs direct access
+------------------------
+
+The VNC, SPICE or serial console network service exposed by QEMU can be made
+to listen on either a local-only UNIX domain socket, or a remotely accessible
+TCP socket. These choices in QEMU configuration, lead to a number of different
+options for how a tenant user can be allowed to connect.
+
+If the QEMU network service listens on a UNIX domain socket, then there will
+need to be a proxy service running on each compute node that exposes the
+service over TCP for remote access. There are several options for proxying, but
+the common approach is to use either a traditional HTTP CONNECT based proxy, or
+a HTTP websockets encoded proxy. In such a scenario, the compute host will only
+need to expose a single TCP port, no matter how many virtual machines are
+running. If tenant users cannot access IP addresses in the cluster network, then
+the POD running the proxy on each host will need an associated service IP
+address, to which tenant users will connect.
+
+If the QEMU network service listens on a TCP socket, then each virtual machine
+running on a compute host will be listening on the same IP address, but using a
```

Comment:
Currently the console is part of virt-handler and virt-api. In contrast to SPICE and VNC it is something which can't be removed. I kind of like this, because it is a very simple mechanism to talk with the VM, like `docker exec`. Although it is just an uninterpreted binary stream, it can be used in telnet fashion very easily and works for a lot of VMs out of the box. Having a guest agent too, which provides an opinionated stream sounds also good, but it would be great if you don't need to alter your VM to get this easy access method. I would also see it as something which does not need an external proxy. I think only in cases where you want to transfer a lot of data through the console endpoint, an external proxy would be required.

For SPICE/VNC (and maybe console if that is really needed) either a dedicated websocket proxy, or a TCP proxy sounds good.